url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/Compare/DataOfPoint.java

```diff
@@ -0,0 +1,198 @@
+import utility.security.SecurityUtility;
+
+public class DataOfPoint{
+	public int pid;
+	public Point p;
+	public Gfunction gf_x, gf_y;
+	public Long[] delaunayIds = null;
+	public String signature;
+	
+	public int getPointId(){
+		return pid;
+	}
+	
+	public DataOfPoint(){
+		p = new Point();
+		delaunayIds = new Long[0];
+		gf_x = new Gfunction();
+		gf_y = new Gfunction();
+	}
+	
+	public DataOfPoint(int id, Point _p, Long[] _ids){
+		pid = id;
+		p = _p;
+		delaunayIds = _ids;
+		gf_x = new Gfunction(p.x, 16);
+		gf_y = new Gfunction(p.y, 16); 
+	}
+	
+	public DataOfPoint(Point _p, Gfunction _gf_x, Gfunction _gf_y){
+		p = _p;
+		gf_x = _gf_x;
+		gf_y = _gf_y;
+	}
+	
+	public DataOfPoint(byte[] buf){
+		this();
+		try {
+			loadFromBytes(buf);
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+	}
+	
+	public byte[] writeToBytes() throws IOException{
+		ByteArrayOutputStream baos = new ByteArrayOutputStream();
+		DataOutputStream dos = new DataOutputStream(baos);
+		dos.writeInt(pid);
+		p.writeToFile(dos);
+		dos.writeInt(delaunayIds.length);
+		for(int i = 0 ; i < delaunayIds.length; i++){
+			dos.writeLong(delaunayIds[i]);
+		}		
+		DataIO.writeString(dos, signature);
+		gf_x.writeToFile(dos);
+		gf_y.writeToFile(dos);
+		return baos.toByteArray();
+	}
+	
+	public void load(DataInputStream dis){
+		try {
+			pid = dis.readInt();
+		} catch (IOException e1) {
+			// TODO Auto-generated catch block
+			e1.printStackTrace();
+		}
+		p.readFromFile(dis);
+		int len;
+		try {
+			len = dis.readInt();
+			delaunayIds = new Long[len];
+			for(int i = 0 ; i < len; i++){
+				delaunayIds[i] = dis.readLong();
+			}
+			signature = DataIO.readString(dis);
+			gf_x.readFromFile(dis);
+			gf_y.readFromFile(dis);
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		
+	}
+	
+	public void loadFromBytes(byte[] buffer) throws IOException{
+		DataInputStream dis = new DataInputStream(new ByteArrayInputStream(buffer));
+		pid = dis.readInt();
+		p.readFromFile(dis);
+		int len = dis.readInt();
+		delaunayIds = new Long[len];
+		for(int i = 0 ; i < len; i++){
+			delaunayIds[i] = dis.readLong();
+		}
+		signature = DataIO.readString(dis);
+		gf_x.readFromFile(dis);
+		gf_y.readFromFile(dis);
+		dis.close();
+	}
+	
+	public String getDigest(){
+		String hashValueOfNei = SecurityUtility.computeGeneralHashValue(delaunayIds);
+		return SecurityUtility.computeGeneralHashValue(new String[]{
+				new Integer(pid).toString(),
+				p.getDigest(),
+				hashValueOfNei
+		});
+	}
+	
+	public String getSignature(){
+		return signature;
+	}
+	
+	public void signWithRSA(RSA rsa){
+		signature = rsa.encrypt(getDigest());
+	}
+	
+	/**
+	 * This is for test.
+	 * @throws IOException 
+	 * */
+	public static void main(String[] args) throws IOException{
+		File file = new File("DataOfPoint.testfile");
+		DataOutputStream dos = new DataOutputStream(new FileOutputStream(file));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *dos*. The resource is referenced at lines 141, 147, 151, 152. The resource is closed at line 153. There are other execution paths that do not contain closure statements, for example, when DataOfPoint.writeToBytes throws an exception. Ensure *dos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).